kernel/src/madt.c

```diff
@@ -0,0 +1,103 @@
+#include "madt.h"
+
+#include "acpi.h"
+#include "kassert.h"
+#include "memory/paging.h"
+#include "io_apic.h"
+
+// Conversion table from 8259 IRQs to GSI
+GSIOverrideEntry g_gsi_override_table[MAX_8259_IRQ_COUNT];
+
+// References to the MADT's IOAPIC entries, uses g_ioapic_count
+IOAPICInfo g_found_ioapics[MAX_IOAPIC_COUNT];
+uint64_t g_ioapic_count = 0;
+
+// References to the MADT's Local APIC entries, uses g_lapic_count
+LAPICInfo* g_found_lapics[MAX_LAPIC_COUNT];
+uint64_t g_lapic_count = 0;
+
+// Local APIC's physical address
+PhysicalAddress g_lapic_physical_address = 0xFEE00000UL;
+
+void parse_madt() {
+    void* madt = find_table("APIC");
+    const ACPISDTHeader* sdt_header = (const ACPISDTHeader*)madt;
+
+    { // Validate MADT
+        bool madt_valid = sdt_is_valid(sdt_header, "APIC");
+        KERNEL_ASSERT(madt_valid, "INVALID MADT");
+    }
+
+    // Fill Override table to have a corrent redirection when GSI = 8259 IRQ.
+    for (int i = 0; i < MAX_8259_IRQ_COUNT; i++) g_gsi_override_table[i].gsi = i;
+
+    const void* madt_end = madt + sdt_header->length;
+
+    // Offset for table fields are 0x2c
+    madt += 0x2c;
+
+    while (madt < madt_end) {
+        const uint8_t type = *(uint8_t*)(madt + 0x0);
+        const uint8_t length = *(uint8_t*)(madt + 0x1);
```

Comment:
Use packed structs instead of manual offsets, it's much more clear what's happening that way.